core/amber/src/test/scala/edu/uci/ics/amber/engine/architecture/scheduling/WorkflowPipelinedRegionsSpec.scala

```diff
@@ -0,0 +1,159 @@
+          OperatorPort(joinOpDesc.operatorID, 0)
+        ),
+        OperatorLink(
+          OperatorPort(headerlessCsvOpDesc2.operatorID, 0),
+          OperatorPort(joinOpDesc.operatorID, 1)
+        ),
+        OperatorLink(
+          OperatorPort(joinOpDesc.operatorID, 0),
+          OperatorPort(sink.operatorID, 0)
+        )
+      )
+    )
+    val pipelinedRegions = new WorkflowPipelinedRegions(workflow)
+    assert(pipelinedRegions.pipelinedRegionsDAG.vertexSet().size == 2)
+
+    var buildRegion: PipelinedRegion = null
+    pipelinedRegions.pipelinedRegionsDAG
+      .vertexSet()
+      .forEach(p =>
+        if (
+          p.getOperators()
+            .contains(
+              OperatorIdentity(workflow.getWorkflowId().id, headerlessCsvOpDesc1.operatorID)
+            )
+        ) {
+          buildRegion = p
+        }
+      )
+
+    var probeRegion: PipelinedRegion = null
+    pipelinedRegions.pipelinedRegionsDAG
+      .vertexSet()
+      .forEach(p =>
+        if (
+          p.getOperators()
+            .contains(
+              OperatorIdentity(workflow.getWorkflowId().id, headerlessCsvOpDesc2.operatorID)
+            )
+        ) {
+          probeRegion = p
+        }
+      )
+
+    assert(pipelinedRegions.pipelinedRegionsDAG.getAncestors(probeRegion).size() == 1)
+    assert(pipelinedRegions.pipelinedRegionsDAG.getAncestors(probeRegion).contains(buildRegion))
+    assert(buildRegion.blockingDowstreamOperatorsInOtherRegions.size == 1)
+    assert(
+      buildRegion.blockingDowstreamOperatorsInOtherRegions.contains(
+        OperatorIdentity(workflow.getWorkflowId().id, joinOpDesc.operatorID)
+      )
+    )
+  }
+
+  "Pipelined Regions" should "correctly find regions in csv->->filter->join->sink workflow" in {
+    val headerlessCsvOpDesc1 = TestOperators.headerlessSmallCsvScanOpDesc()
+    val keywordOpDesc = TestOperators.keywordSearchOpDesc("column-1", "Asia")
+    val joinOpDesc = TestOperators.joinOpDesc("column-1", "column-1")
+    val sink = TestOperators.sinkOpDesc()
+    val workflow = buildWorkflow(
+      mutable.MutableList[OperatorDescriptor](
+        headerlessCsvOpDesc1,
+        keywordOpDesc,
+        joinOpDesc,
+        sink
+      ),
+      mutable.MutableList[OperatorLink](
+        OperatorLink(
+          OperatorPort(headerlessCsvOpDesc1.operatorID, 0),
+          OperatorPort(joinOpDesc.operatorID, 0)
+        ),
+        OperatorLink(
+          OperatorPort(headerlessCsvOpDesc1.operatorID, 0),
+          OperatorPort(keywordOpDesc.operatorID, 0)
+        ),
+        OperatorLink(
+          OperatorPort(keywordOpDesc.operatorID, 0),
+          OperatorPort(joinOpDesc.operatorID, 1)
+        ),
+        OperatorLink(
+          OperatorPort(joinOpDesc.operatorID, 0),
+          OperatorPort(sink.operatorID, 0)
+        )
+      )
+    )
+    val pipelinedRegions = new WorkflowPipelinedRegions(workflow)
+    assert(pipelinedRegions.pipelinedRegionsDAG.vertexSet().size == 1)
+  }
```

Comment:
Maybe add a test case based on this use case:

basically the first and the third join operator share the same build operator

<img width="965" alt="Screen Shot 2022-07-12 at 6 38 40 PM 1" src="https://user-images.githubusercontent.com/12578068/178629931-a5c56604-1286-47f6-8a57-33c49b8ae6b6.png">

Reply:
done